src/pam/main.cc

```diff
@@ -0,0 +1,394 @@
+  // Will contain the responses from PAM functions
+  int pam_res = PAM_IGNORE;
+
+  // Try to get the conversation function and error out if we can't
+  if ((pam_res = pam_get_item(pamh, PAM_CONV, (const void **)&conv)) !=
+      PAM_SUCCESS) {
+    syslog(LOG_ERR, "Failed to acquire conversation");
+    return pam_res;
+  }
+
+  // Wrap the PAM conversation function in our own, easier function
+  auto conv_function =
+      bind(send_message, conv->conv, placeholders::_1, placeholders::_2);
+
+  // Error out if we could not ready the config file
+  if (reader.ParseError() < 0) {
+    syslog(LOG_ERR, "Failed to parse the configuration file");
+    return PAM_SYSTEM_ERR;
+  }
+
+  // Stop executing if Howdy has been disabled in the config
+  if (reader.GetBoolean("core", "disabled", false)) {
+    syslog(LOG_INFO, "Skipped authentication, Howdy is disabled");
+    return PAM_AUTHINFO_UNAVAIL;
+  }
+
+  // Stop if we're in a remote shell and configured to exit
+  if (reader.GetBoolean("core", "ignore_ssh", true)) {
+    if (getenv("SSH_CONNECTION") != nullptr ||
+        getenv("SSH_CLIENT") != nullptr || getenv("SSHD_OPTS") != nullptr) {
+      syslog(LOG_INFO, "Skipped authentication, SSH session detected");
+      return PAM_AUTHINFO_UNAVAIL;
+    }
+  }
+
+  // Try to detect the laptop lid state and stop if it's closed
+  if (reader.GetBoolean("core", "ignore_closed_lid", true)) {
+    glob_t glob_result{};
+
+    // Get any files containing lid state
+    int return_value =
+        glob("/proc/acpi/button/lid/*/state", 0, nullptr, &glob_result);
+
+    // TODO: We ignore the result
+    if (return_value != 0) {
+      globfree(&glob_result);
+    }
+
+    for (size_t i = 0; i < glob_result.gl_pathc; i++) {
+      ifstream file(string(glob_result.gl_pathv[i]));
+      string lid_state;
+      getline(file, lid_state, (char)file.eof());
+
+      if (lid_state.find("closed") != std::string::npos) {
+        globfree(&glob_result);
+
+        syslog(LOG_INFO, "Skipped authentication, closed lid detected");
+        return PAM_AUTHINFO_UNAVAIL;
+      }
+    }
+
+    globfree(&glob_result);
+  }
+
+  // If enabled, send a notice to the user that facial login is being attempted
+  if (reader.GetBoolean("core", "detection_notice", false)) {
+    if ((pam_res = conv_function(
+             PAM_TEXT_INFO,
+             dgettext("pam", "Attempting facial authentication"))) !=
+        PAM_SUCCESS) {
+      syslog(LOG_ERR, "Failed to send detection notice");
+    }
+  }
+
+  // Get the username from PAM, needed to match correct face model
+  char *user_ptr = nullptr;
+  if ((pam_res = pam_get_user(pamh, (const char **)&user_ptr, nullptr)) !=
+      PAM_SUCCESS) {
+    syslog(LOG_ERR, "Failed to get username");
+    return pam_res;
+  }
+
+  posix_spawn_file_actions_t file_actions;
+  posix_spawn_file_actions_init(&file_actions);
+  // We close standard descriptors for the child
+  posix_spawn_file_actions_addclose(&file_actions, STDOUT_FILENO);
```

Comment:
Is there a particular reason for closing stdout/stderr? This is actually causing a crash when we have `end_report = true` set in the config and the python script tries to print the diagnostics. It crashes because stdout is not writable, meaning that the pam module assumes that the auth failed.

Reply:
IIRC, it was indeed because the script was crashing when stdin was kept open. For stout and stderr, I don't remember a particular reason to close them, maybe just to take (too much) precautions.